### PR TITLE
WIP: Jsonschema support

### PR DESCRIPTION
**Pull request by nfwvogt**

### CODE REVIEW

File: struqture/src/spins/spin_operator.rs [hunk 1]
```diff
@@ -52,16 +52,37 @@ use std::ops;
 /// ```
 ///
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+// #[cfg_attr(feature="schema", derive(schemars::JsonSchema))]
```
Comment (kbarkhqs) [inline struqture/src/spins/spin_operator.rs:55]: This should be deleted

File: struqture/tests/spins/pauli_product.rs [hunk 2]
```diff
@@ -541,3 +541,18 @@ fn test_singlespinoperator_product() {
         assert_eq!(test_mat, direct_matrix_multiplication)
     }
 }
+
+#[cfg(feature = "json_schema")]
+#[cfg(feature = "json_schema")]
+#[cfg(feature = "json_schema")]
```
Comment (kbarkhqs) [inline struqture/tests/spins/pauli_product.rs:547]: Two of these can be removed

File: struqture/tests/spins/pauli_product.rs [hunk 3]
```diff
@@ -541,3 +541,18 @@ fn test_singlespinoperator_product() {
...
+#[test]
+fn test_pauli_product_schema() {
+    let pp = PauliProduct::new();
+    let schema = schemars::schema_for!(PauliProduct);
+    let schema_checker = jsonschema::JSONSchema::compile(&serde_json::to_value(&schema).unwrap())
+        .expect("schema is valid");
+    let value = serde_json::to_value(pp).unwrap();
+    let validation = schema_checker.validate(&value);
+    assert!(validation.is_ok());
+    println!("{}", serde_json::to_string_pretty(&schema).unwrap());
```
Comment (kbarkhqs) [inline struqture/tests/spins/pauli_product.rs:557]: This line can be removed

File: struqture/tests/spins/spin_hamiltonian.rs [hunk 4]
```diff
@@ -810,3 +810,20 @@ fn sparse_lindblad_entries() {
...
+    let mut op = SpinHamiltonian::new();
+    op.set(PauliProduct::new().x(0), 1.0.into()).unwrap();
+    op.set(PauliProduct::new().y(1).z(2), "val".into()).unwrap();
+    let schema = schemars::schema_for!(SpinHamiltonian);
+    let schema_checker = jsonschema::JSONSchema::compile(&serde_json::to_value(&schema).unwrap())
+        .expect("schema is valid");
+    let value = serde_json::to_value(&op).unwrap();
+    let validation = schema_checker.validate(&value);
+    println!("{}", serde_json::to_string_pretty(&schema).unwrap());
+    println!("{}", serde_json::to_string_pretty(&op).unwrap());
```
Comment (kbarkhqs) [inline struqture/tests/spins/spin_hamiltonian.rs:826]: These lines can be removed or turned into assertions

File: struqture/src/bosons/bosonic_system.rs [hunk 5]
```diff
@@ -53,6 +53,8 @@ pub struct BosonSystem {
     pub(crate) operator: BosonOperator,
```
Comment (kbarkhqs) [inline struqture/src/bosons/bosonic_system.rs:53]: This should also contain the 
#[cfg_attr(feature = "json_schema", derive(schemars::JsonSchema))]
#[cfg_attr(feature = "json_schema", schemars(deny_unknown_fields))]

lines

**Review by kbarkhqs — APPROVED** [review]

File: struqture/src/bosons/bosonic_system.rs (hunk 5 above)
Comment (nfwvogt) [reply struqture/src/bosons/bosonic_system.rs:53]: Fixed

File: struqture/src/mixed_systems/mod.rs [hunk 6]
```diff
@@ -41,6 +41,7 @@ use crate::{
     bosons::BosonIndex, fermions::FermionIndex, ModeIndex, OperateOnDensityMatrix, SpinIndex,
     StruqtureError,
 };
+use mixed_noise_system::TinyVecDef;
```
Comment (kbarkhqs) [inline struqture/src/mixed_systems/mod.rs:44]: This doesn't seem to have propagated to the tests

File: struqture/src/spins/spin_operator.rs (hunk 1 above)
Comment (nfwvogt) [reply struqture/src/spins/spin_operator.rs:55]: Fixed

**Pull request merged**